### ISSUE TITLE
Key are not deleted after TTL expire

### ISSUE THREAD

**Issue by lukass77** [facebook/rocksdb#12116]
### Expected behavior
when using TTL , expected behavior is that the key/value entry will be deleted when TTL is expire

### Actual behavior
Executed a simple program that put key/value entry into rocks db , set TTL to be 10 seconds , stop thread for 20 seconds 
key/value entry is not deleted 

### Question
use case I looking for is to set TTL , and that when TTL expire key/value entry will be deleted automaticly , base only the time entry "sit" in database  , is this supported on Rocks-db ?
can you please provide a configuration sample that will do that ?

### Steps to reproduce the behavior

`package com.rocksdb.sample;

import org.rocksdb.CompactionOptionsFIFO;
import org.rocksdb.Env;
import org.rocksdb.Options;
import org.rocksdb.RocksDB;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

import static org.rocksdb.util.SizeUnit.MB;

public class RocksDbTTLSample {

    private static RocksDB rockdDb = null;

    public static void main(String[] args) {

        try {
            RocksDB.loadLibrary();
            Options options = new Options();
            options.setCreateIfMissing(true);
            options.setEnv(Env.getDefault());

            //fifo related
            CompactionOptionsFIFO compactionOptionsFIFO = new CompactionOptionsFIFO();
            compactionOptionsFIFO.setAllowCompaction(true);
            compactionOptionsFIFO.setMaxTableFilesSize(500 * MB);
            options.setCompactionOptionsFIFO(compactionOptionsFIFO);

            options.setTtl(10L);
            System.out.println("set TTL to 10 seconds");
            options.setPeriodicCompactionSeconds(10);
            options.setMaxBackgroundJobs(4);
            options.setMaxOpenFiles(-1);

            System.out.println("Rocks-db  - done configuration");
            rockdDb = RocksDB.open(options, "/tmp/myRocksDb");
            System.out.println("Rocks-db is started");

            //put key value entry
            Charset charset = StandardCharsets.US_ASCII;
            String key = "myKey";
            String value = "myValue";
            rockdDb.put(key.getBytes(charset), value.getBytes(charset));
            System.out.println("saved key/value in rocks db for key - " + key + " with value " + value);
            System.out.println("waiting for 20 second for apply the TTL and key eviction");
            Thread.sleep(20000);

            //get value
            byte[] valueOfMyKey = rockdDb.get(key.getBytes(charset));

            if (valueOfMyKey.length > 0) {
                System.out.println("found value for key - " + key);
                System.out.println("TTL not work as expected as key/value entry is not evicted from db");
            }
        } catch (Exception e) {
            System.out.println("error rocks db");
            e.printStackTrace();
            throw new RuntimeException(e);
        } finally {
            rockdDb.close();
        }
    }
}`

**Comment by lukass77** [comment]
[RocksDbSample.zip](https://github.com/facebook/rocksdb/files/13532699/RocksDbSample.zip)
attached java project , rocks-db version is 8.8.1

**Comment by lukass77** [comment]
it appear that only after calling programtilcy to 
  rockdDb.compactRange(); , after this line -  Thread.sleep(20000);
  
  to some up - TTL is not taking place automatilcy but only upon on calling rockdDb.compactRange()

**Comment by alanpaxton** [comment]
Hi @lukass77 the wiki (https://github.com/facebook/rocksdb/wiki/Time-to-Live) states, re TTL behaviour

"Expired TTL values are deleted in compaction only"
"Get/Iterator may return expired entries(compaction not run on them yet)"

That's kind of what I would expect; it's far less resource intensive to just check for TTL on a compaction pass than for instance to actively run a timer queue.

If you are still seeing (key,value)s visible after compaction, then let us know and we can investigate further.

**Comment by lukass77** [comment]
Thanks @alanpaxton , 
FYI @adamretter 

I understand your point , so entry will be deleted on compaction .
now what I am facing is that compaction is not executed automatically as a background thread but I need to execute an api call rockdDb.compactRange() for do compaction which will cause the TTL expired entries to be deleted .

Question
can we have the compaction tune/configure to run automatically as a daemon/background thread and avoid the programtic call to compactRange() ?

you can look on the sample code above I am setting options.setPeriodicCompactionSeconds(10) ,  and then use Thread.sleep(20000) 20 seconds --> but entries not evicted .

can it be there is an issue with options.setPeriodicCompactionSeconds(10) ? 
I will appreciate if you can show what need to be done to have the compaction runs automatically in the backround  on specific time interval .

**Comment by alanpaxton** [comment]
Hi @lukass77 - I am taking a further look at this to see if I can track down a way to understand when periodic compaction will happen. You can `grep -i "compacted to" <database-folder>/LOG` to see reports of compaction runs, which I have forced by writing the same batch of (k,v) pairs repeatedly - I suspect any issue is more along the lines of RocksDB deciding a compaction may not reclaim much space - but I have not convinced it to remove "key" without calling `compactRange()`.

Of course given the semantics, any system that needs to respect TTL exactly will need to check write times itself of any value it reads back.

**Comment by lukass77** [comment]
Hi @alanpaxton - as this issue is closed do you prefer to have a new issue "PeriodicCompaction not working out of the box" 
so it will be more visable on git .

second - you can try my code above  uploaded zip file , I maybe need to tune configuration to make PeriodicCompaction works without the explicit programtic call to compactRange()

question is if PeriodicCompaction in case it do works  actually causing a background execution of compactRange() ?

**Comment by adamretter** [comment]
@lukass77 I was perhaps to hasty - I have reopened this issue.

**Comment by rhubner** [comment]
Hello @lukass77,
I took over this issue from @alanpaxton  and probably found some reasons why your example doesn't work. To make it work require few prerequisites: 

- RocksDB must be open via the `TtlDB` class.
- There must be at least two SST files on disk for competitions. 

Please check my [test example ](https://github.com/evolvedbinary/rocksdb/blob/eb/ttl-compaction/java/src/test/java/org/rocksdb/TTLExpiryTest.java) where I simulated this behaviour. I'm opening a database via `RocksDB `class and `TtlDB`. As you can see, on the test result, on `TtlDB ` in the third iteration, the key is not there. 

At the moment, it looks like there is a small bug in `TtlDB ` with the default Column family. I need to double check and will send PR sooner.

**Comment by lukass77** [comment]
Hi @rhubner , 
your second point - "There must be at least two SST files on disk for competitions." , not sure I understnad why .. but is this a limitation for support TTL on rocks db / how can we make sure we have 2 sst files ?

with regards your test example why we need to do flush after put .. ?
t```
ry(FlushOptions fOptions = new FlushOptions()) {
                    db.flush(fOptions);
                }

```
is this impact if TTL is working or not ? , should I also execute flush on production code ? - as my example main did not use it
I will try to adept my main example and share the result here
Thanks!!!

**Comment by lukass77** [comment]
Hey @rhubner 
I succeed to make my program TTL to work after I added second "put"
with that still few issues , as it seems the TTL is not working out of the box ...

1. for TTL to work I must use db.flush , when db.flush not called TTL not working , do we must call db.flush for TTL to work ?
2. for TTL to work I must use 2 times sleep call once after the first key put and second time after the second key put , not only once after the second key put (what I thought should work ..)
3. use second key to make TTL works , I think that TTL should not be depend on number of keys you have in the database , and this can be fixed , or to be added in documentation as a limitation .

upload my working version , if you comment out one of the sleep calls / or you comment out db.flush TTL is not working
[RocksDbSample.zip](https://github.com/facebook/rocksdb/files/13757223/RocksDbSample.zip)

**Comment by lukass77** [comment]
`
package com.rocksdb.sample;

import org.rocksdb.*;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import static org.rocksdb.util.SizeUnit.MB;


public class RocksDbTTLSample {
    private static final int TTL = 10;

    private static TtlDB rocksdb = null;

    public static void main(String[] args) {

        try {
            Options options = new Options();
            options.setCreateIfMissing(true);
            options.setEnv(Env.getDefault());

            //fifo related
            CompactionOptionsFIFO compactionOptionsFIFO = new CompactionOptionsFIFO();
            compactionOptionsFIFO.setAllowCompaction(true);
            compactionOptionsFIFO.setMaxTableFilesSize(500 * MB);
            options.setCompactionOptionsFIFO(compactionOptionsFIFO);

            options.setTtl(TTL);
            System.out.println("set TTL to :" + TTL + " seconds");
            options.setPeriodicCompactionSeconds(TTL);
            options.setMaxBackgroundJobs(4);
            options.setMaxOpenFiles(-1);

            System.out.println("Rocks-db  - done configuration");
            rocksdb = TtlDB.open(options, "/tmp/myRocksDb", TTL, false);

            System.out.println("Rocks-db is started");
            System.out.println(Arrays.toString(rocksdb.getDefaultColumnFamily().getName()));
            System.out.println(RocksDB.rocksdbVersion().toString());


            //put key-1 value entry
            Charset charset = StandardCharsets.US_ASCII;
            String key = "myKey";
            String value = "myValue";
            rocksdb.put(key.getBytes(charset), value.getBytes(charset));
            rocksdb.flush(new FlushOptions());
            System.out.println("saved key/value in rocks db for key - " + key + " with value " + value);

            System.out.println("waiting for 15 second for apply the TTL and key eviction (key - 1)");
            Thread.sleep(TimeUnit.SECONDS.toMillis(TTL + 5));

            //key-2
            String key2 = "myKey2";
            String value2 = "myValue2";
            rocksdb.put(key2.getBytes(charset), value2.getBytes(charset));
            rocksdb.flush(new FlushOptions());
            System.out.println("saved key/value in rocks db for key - " + key2 + " with value " + value2);

            System.out.println("waiting for 15 second for apply the TTL and key eviction (key - 2)");
            Thread.sleep(TimeUnit.SECONDS.toMillis(TTL + 5));

            boolean keyExists = rocksdb.keyExists(key.getBytes(charset));
            if (!keyExists) {
                System.out.println("TTL works as expected");
            } else {
                System.out.println("TTL not works as expected , key should be removed but it does not!!");
            }

        } catch (Exception e) {
            System.out.println("error rocks db");
            e.printStackTrace();
            throw new RuntimeException(e);
        } finally {
            rocksdb.close();
        }
    }
}
`

**Comment by rhubner** [comment]
Hello @lukass77,
TTL in RocksDB has the opposite guarantee that you need. It guarantees that data will be there **AT LEAST TTL** seconds. Not the opposite.

> why we need to do flush after put

This is related to how RocksDB is working. On write, all data is written(in default configuration) only to the memory and WAL(Write ahead log). With this functionality RocksDB has very low write amplification. Once the memory(memtable) reaches a certain size, the whole memtable is written to disk to Static Sorted Table(SST). When you have multiple SST files, they are then compacted to one file and this is when keys with expired TTL are removed.

RocksDB is quite a low level library. I suggest checking [RocksDB-Overview](https://github.com/facebook/rocksdb/wiki/RocksDB-Overview). This should help you to understand how TTL is implemented in RocksDB.

**Comment by scglenn** [comment]
@rhubner 
Are there any example tests for auto-compaction removing keys after TTL? Would a test just loop and wait for the key to be deleted? I have found no examples like this, only examples of executing manual compactions.

**Comment by rhubner** [comment]
Hello @scglenn, 

I'm primary familiar with Java part of RocksDB and I'm not aware of any type of this test. Problem with these types of tests is that they can be unpredictable and generate failure in CI/CD.

When you create this type of test, you can tweak some of the [compaction setting](https://github.com/facebook/rocksdb/wiki/Leveled-Compaction) to achieve sooner compaction. This should speedup test execution and reduce the amount of data needed to write to RocksDB before automatic compaction is triggered.
